analyzers/src/SonarAnalyzer.CSharp/Rules/NonDerivedPrivateClassesShouldBeSealed.cs

```diff
@@ -24,48 +24,78 @@ namespace SonarAnalyzer.Rules.CSharp
-                    && !HasVirtualMembers(typeDeclarationSyntax))
+                var declarations = c.Tree.GetCompilationUnitRoot().DescendantNodes(x => x.IsAnyKind(KindsToBeDescended))
+                     .Where(x => x.IsAnyKind(SyntaxKind.ClassDeclaration, SyntaxKindEx.RecordClassDeclaration))
+                     .Select(x => (TypeDeclarationSyntax)x);
+
+                var model = new Lazy<SemanticModel>(() => c.Compilation.GetSemanticModel(c.Tree));
+                var symbols = new Lazy<List<INamedTypeSymbol>>(() => declarations.Select(x => model.Value.GetDeclaredSymbol(x)).ToList());
+
+                foreach (var declaration in declarations)
                 {
-                    var nestedPrivateTypeInfo = (INamedTypeSymbol)c.SemanticModel.GetDeclaredSymbol(c.Node);
-                    if (!IsPrivateTypeInherited(nestedPrivateTypeInfo))
+                    if (IsNotSealed(declaration)
+                        && !HasVirtualMembers(declaration)
+                        && !IsPossiblyDerived(declaration, model, symbols, out var modifierDescription))
                     {
-                        c.ReportIssue(Diagnostic.Create(Rule, typeDeclarationSyntax.Identifier.GetLocation()));
+                        var typeDescription = declaration.IsKind(SyntaxKind.ClassDeclaration) ? "classes" : "records";
+                        c.ReportIssue(Diagnostic.Create(Rule, declaration.Identifier.GetLocation(), modifierDescription, typeDescription));
                     }
                 }
-            },
-            SyntaxKind.ClassDeclaration,
-            SyntaxKindEx.RecordClassDeclaration);
+            });
 
-        private static bool HasVirtualMembers(TypeDeclarationSyntax typeDeclaration)
-        {
-            var classMembers = typeDeclaration.Members;
-            return classMembers.OfType<MethodDeclarationSyntax>().Any(member => member.Modifiers.Any(SyntaxKind.VirtualKeyword))
-                   || classMembers.OfType<PropertyDeclarationSyntax>().Any(member => member.Modifiers.Any(SyntaxKind.VirtualKeyword))
-                   || classMembers.OfType<IndexerDeclarationSyntax>().Any(member => member.Modifiers.Any(SyntaxKind.VirtualKeyword))
-                   || classMembers.OfType<EventDeclarationSyntax>().Any(member => member.Modifiers.Any(SyntaxKind.VirtualKeyword));
-        }
+        private static bool HasVirtualMembers(TypeDeclarationSyntax typeDeclaration) =>
+            typeDeclaration.Members
+                .Where(member => member.IsAnyKind(PossiblyVirtualKinds))
+                .Any(member => member.Modifiers().Any(SyntaxKind.VirtualKeyword));
 
-        private static bool IsPrivateButNotSealedType(TypeDeclarationSyntax typeDeclaration) =>
-            typeDeclaration.Modifiers.Any(SyntaxKind.PrivateKeyword)
-            && !typeDeclaration.Modifiers.Any(SyntaxKind.StaticKeyword)
+        private static bool IsNotSealed(TypeDeclarationSyntax typeDeclaration) =>
```

Comment:
I would suggest avoiding the negation (`IsSealed`) and inverting the implementation